### PR TITLE
[Impeller] Color matrix color filter implementation.

### PR DESCRIPTION
**Pull request by betrevisan**
The changes proposed in this PR implement the color matrix color filter in Impeller alongside accompanying tests. One of the added tests allows developers to play around with the behavior of the color filter as they change the values of the matrix.

The PR also fixes https://github.com/flutter/flutter/issues/107358.

This PR was developed on top of the changes proposed earlier in https://github.com/flutter/engine/pull/34483.

Below is a Demo of what the playground looks like for this filter.
Screen Shot 2022-08-10 at 9 02 46 AM

Screen Shot 2022-08-10 at 9 03 00 AM

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: impeller/entity/entity_unittests.cc [hunk 1]
```diff
@@ -1393,5 +1393,98 @@ TEST_P(EntityTest, RRectShadowTest) {
...
+      ImGui::SetNextWindowPos({260, 600});
+    }
+
+    // UI state.
+    static FilterContents::ColorMatrix color_matrix = {
+        1, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0};
+
+    // Define the ImGui
+    ImGui::Begin("Color Matrix");
+    char label[4] = "##1";
```
Comment (iskakaushik) [inline impeller/entity/entity_unittests.cc:1463]: nit: use `std::string label = "##1"; label.c_str(); // next line`, makes it easier to change in the future.

**Review by iskakaushik — APPROVED** [review]
LGTM

File: impeller/entity/shaders/color_matrix_color_filter.frag [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+void main() {
+  vec4 input_color = texture(input_texture, v_position);
+
+  // unpremultiply first, as filter inputs are premultiplied.
+  vec4 color = IPUnpremultiply(input_color);
+
+  color = frag_info.color_m * color + frag_info.color_v;
+
+  // color = clamp(color);
```
Comment (jonahwilliams) [inline impeller/entity/shaders/color_matrix_color_filter.frag:46]: drive by nit: commented out code

Comment (iskakaushik) [reply impeller/entity/shaders/color_matrix_color_filter.frag:46]: @betrevisan  could you please fix this :-)

Comment (betrevisan) [reply impeller/entity/shaders/color_matrix_color_filter.frag]: Yes, just updated it. Thanks for catching that :)

File: impeller/entity/entity_unittests.cc [hunk 3]
```diff
@@ -1393,5 +1393,99 @@ TEST_P(EntityTest, RRectShadowTest) {
...
+TEST_P(EntityTest, ColorMatrixFilterCoverageIsCorrect) {
+  // Set up a simple color background.
+  auto fill = std::make_shared<SolidColorContents>();
+  fill->SetPath(
+      PathBuilder{}.AddRect(Rect::MakeXYWH(0, 0, 300, 400)).TakePath());
+  fill->SetColor(Color::Coral());
+
+  // Set the color matrix filter.
+  FilterContents::ColorMatrix matrix = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+                                        1, 1, 1, 1, 1, 1, 1, 1, 1, 1};
```
Comment (bdero) [inline impeller/entity/entity_unittests.cc:1405]: Nit: For these and all of the other matrices, maybe lay these out in a 5x4 grid for easy reading? Often we use comments to coerce a formatter in situations like this. For example
```
FilterContents::ColorMatrix matrix = {
    1, 1, 1, 1, 1, //
    1, 1, 1, 1, 1, //
    1, 1, 1, 1, 1, //
    1, 1, 1, 1, 1, //
};
```

**Review by bdero — APPROVED** [review]
LGTM.

I haven't actually tried the playground -- a screenshot or video of the test would be nice to see (but by no means required). :)

File: impeller/entity/entity_unittests.cc (hunk 3 above)
Comment (betrevisan) [reply impeller/entity/entity_unittests.cc]: Nice! I edited the matrices now.

**Comment by iskakaushik** [comment]
@betrevisan could you please fix the formatting errors?

**Comment by betrevisan** [comment]
> @betrevisan could you please fix the formatting errors?

**Pull request merged**